workers/unity/Packages/com.improbable.gdk.core/Worker/ConnectionHandlers/MultithreadedSpatialOSConnectionHandler/ThreadedSerializationHandler.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Improbable.Gdk.Core
+{
+    internal class ThreadedSerializationHandler : IDisposable
+    {
+        private readonly CommandMetaDataManager commandMetaDataManager;
+
+        private readonly BlockingCollection<MessagesToSend> messagesToSerialize;
+        private readonly ConcurrentPool<MessagesToSend> messagesToSendPool;
+
+        private readonly ConcurrentQueue<SerializedMessagesToSend> serializedMessages;
+        private readonly ConcurrentPool<SerializedMessagesToSend> serializedMessagesToSendPool;
+
+        private readonly Thread serialisationThread;
+
+        public ThreadedSerializationHandler(CommandMetaDataManager commandMetaDataManager)
+        {
+            this.commandMetaDataManager = commandMetaDataManager;
+
+            messagesToSerialize = new BlockingCollection<MessagesToSend>(new ConcurrentQueue<MessagesToSend>());
+            serializedMessagesToSendPool = new ConcurrentPool<SerializedMessagesToSend>();
+            messagesToSendPool = new ConcurrentPool<MessagesToSend>();
+            serializedMessages = new ConcurrentQueue<SerializedMessagesToSend>();
+
+            serialisationThread = new Thread(SerializeMessages);
+            serialisationThread.Start();
+        }
+
+        public void EnqueueMessagesToSend(MessagesToSend messages)
+        {
+            messagesToSerialize.Add(messages);
+        }
+
+        public bool TryDequeueSerializedMessages(out SerializedMessagesToSend serializedMessagesToSend)
+        {
+            return serializedMessages.TryDequeue(out serializedMessagesToSend);
+        }
+
+        public MessagesToSend GetMessagesToSendContainer()
+        {
+            return messagesToSendPool.Rent();
+        }
+
+        public void ReturnSerializedMessageContainer(SerializedMessagesToSend serializedMessagesToSend)
+        {
+            serializedMessagesToSendPool.Return(serializedMessagesToSend);
+        }
+
+        private void SerializeMessages()
+        {
+            // Block until a new message is available to serialize.
+            while (messagesToSerialize.TryTake(out var messagesToSend, -1))
+            {
+                // If adding is complete then there is no reason to deserialize the message.
+                if (messagesToSerialize.IsAddingCompleted)
```

Comment:
The name of this is slightly odd - afaict its just a signal to the thread that its time to shut down.

Reply:
The `IsAddingComplete`? That's part of `BlockingCollection<>` rather than something I made. We can add another explicit signal instead but then we have to lock around it.

Reply:
It stops you being able to add more messages to the queue and cause another flag to be set when it's then emptied.

Reply:
In which case - forgot I said that 😬